### PR TITLE
add JVMTI callback/hook for virtualThread

### PR DESCRIPTION
**Pull request by fengxue-IS**
Related: #15183 

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

### CODE REVIEW

File: runtime/jvmti/jvmtiCapability.c [hunk 1]
```diff
@@ -187,6 +187,14 @@ jvmtiGetPotentialCapabilities(jvmtiEnv* env, jvmtiCapabilities* capabilities_ptr
 	}
 #endif /* JAVA_SPEC_VERSION >= 11 */
 
+#if JAVA_SPEC_VERSION >= 19
+	if (isEventHookable(j9env, JVMTI_EVENT_VIRTUAL_THREAD_START) &&
+		isEventHookable(j9env, JVMTI_EVENT_VIRTUAL_THREAD_END)
```
Comment (babsingh) [inline runtime/jvmti/jvmtiCapability.c:192]:
```suggestion
	if (isEventHookable(j9env, JVMTI_EVENT_VIRTUAL_THREAD_START)
	&& isEventHookable(j9env, JVMTI_EVENT_VIRTUAL_THREAD_END)
```

File: runtime/jvmti/jvmtiHook.c [hunk 2]
```diff
@@ -145,6 +145,10 @@ static void jvmtiHookGetEnv (J9HookInterface** hook, UDATA eventNum, void* event
 static void jvmtiHookVmDumpStart (J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
 static void jvmtiHookVmDumpEnd (J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
 static UDATA methodExists(J9Class * methodClass, U_8 * nameData, UDATA nameLength, J9UTF8 * signature);
+#if JAVA_SPEC_VERSION >= 19
+static void jvmtiHookVirtualThreadStarted (J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
+static void jvmtiHookVirtualThreadEnd (J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData);
```
Comment (babsingh) [inline runtime/jvmti/jvmtiHook.c:150]:
```suggestion
static void jvmtiHookVirtualThreadStarted (J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
static void jvmtiHookVirtualThreadEnd (J9HookInterface **hook, UDATA eventNum, void *eventData, void *userData);
```

File: runtime/jvmti/jvmtiHook.c [hunk 3]
```diff
@@ -337,6 +341,63 @@ jvmtiHookMethodEnter(J9HookInterface** hook, UDATA eventNum, void* eventData, vo
...
+	ENSURE_EVENT_PHASE_START_OR_LIVE(jvmtiHookVirtualThreadStarted, j9env);
+
+	/* Call the event callback */
+
+	if (callback != NULL) {
+		jthread threadRef;
+		UDATA hadVMAccess;
+		UDATA javaOffloadOldState = 0;
+
+		if (prepareForEvent(j9env, currentThread, currentThread, JVMTI_EVENT_VIRTUAL_THREAD_START, &threadRef, &hadVMAccess, FALSE, 0, &javaOffloadOldState)) {
```
Comment (babsingh) [inline runtime/jvmti/jvmtiHook.c:364]: code-lines more than 100 chars should be spread across multiple lines. this should be applied to all places where code-line exceeds 100 chars.
```suggestion
		if (prepareForEvent(
				j9env, currentThread, currentThread, JVMTI_EVENT_VIRTUAL_THREAD_START,
				&threadRef, &hadVMAccess, FALSE, 0, &javaOffloadOldState)
		) {
```

File: runtime/jvmti/jvmtiHook.c [hunk 4]
```diff
@@ -337,6 +341,63 @@ jvmtiHookMethodEnter(J9HookInterface** hook, UDATA eventNum, void* eventData, vo
...
+			callback((jvmtiEnv *) j9env, (JNIEnv *) currentThread, threadRef);
+			finishedEvent(currentThread, JVMTI_EVENT_VIRTUAL_THREAD_START, hadVMAccess, javaOffloadOldState);
+		}
+	}
+
+	TRACE_JVMTI_EVENT_RETURN(jvmtiHookVirtualThreadStarted);
+}
+
+static void
+jvmtiHookVirtualThreadEnd(J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData)
```
Comment (babsingh) [inline runtime/jvmti/jvmtiHook.c:374]: all comments for `jvmtiHookVirtualThreadStarted` also apply to `jvmtiHookVirtualThreadEnd`.

File: runtime/jvmti/jvmtiHook.c [hunk 5]
```diff
@@ -337,6 +341,63 @@ jvmtiHookMethodEnter(J9HookInterface** hook, UDATA eventNum, void* eventData, vo
...
+
+	/* Call the event callback */
+
+	if (callback != NULL) {
+		jthread threadRef;
+		UDATA hadVMAccess;
+		UDATA javaOffloadOldState = 0;
+
+		if (prepareForEvent(j9env, currentThread, currentThread, JVMTI_EVENT_VIRTUAL_THREAD_START, &threadRef, &hadVMAccess, FALSE, 0, &javaOffloadOldState)) {
+			callback((jvmtiEnv *) j9env, (JNIEnv *) currentThread, threadRef);
```
Comment (babsingh) [inline runtime/jvmti/jvmtiHook.c:365]:
```suggestion
			callback((jvmtiEnv *)j9env, (JNIEnv *)currentThread, threadRef);
```

File: runtime/jvmti/jvmtiHook.c (hunk 2 above)
Comment (fengxue-IS) [reply runtime/jvmti/jvmtiHook.c:150]: I am leaning towards keeping the existing format here, or else we could just change all the function headers to match standard

**Review by babsingh — APPROVED** [review]
LGTM, approving. 

I would suggest strengthening the content in the commit message and PR description by highlighting the specifics for which support is introduced:
- `jvmtiHookVirtualThreadStarted`
- `jvmtiHookVirtualThreadEnd`
- `can_support_virtual_threads`

**Review by tajila — APPROVED** [review]

File: runtime/include/jvmti.h.m4 [hunk 6]
```diff
@@ -744,7 +744,10 @@ ifelse(eval(JAVA_SPEC_VERSION >= 11), 1, [	JVMTI_EVENT_SAMPLED_OBJECT_ALLOC = 86
 ifelse(eval(JAVA_SPEC_VERSION >= 19), 1, [	JVMTI_EVENT_VIRTUAL_THREAD_START = 87,], [dnl])
 ifelse(eval(JAVA_SPEC_VERSION >= 19), 1, [	JVMTI_EVENT_VIRTUAL_THREAD_END = 88,], [dnl])
 
-	JVMTI_MAX_EVENT_TYPE_VAL = 86,
+ifelse(eval(JAVA_SPEC_VERSION >= 19), 1, [	JVMTI_MAX_EVENT_TYPE_VAL = 88,],[
+ifelse(eval(JAVA_SPEC_VERSION >= 11), 1, [	JVMTI_MAX_EVENT_TYPE_VAL = 86,],[
+	JVMTI_MAX_EVENT_TYPE_VAL = 84,
+])])
```
Comment (tajila) [inline runtime/include/jvmti.h.m4:750]: do we need the trailing comma `JVMTI_MAX_EVENT_TYPE_VAL = 84,` ?

Also in other places `])])` is not on a separate line

Comment (tajila) [reply runtime/include/jvmti.h.m4:750]: there is an example in line 60 on how this is done

Comment (fengxue-IS) [reply runtime/include/jvmti.h.m4]: the comma separates this line with the next line of `jvmtiEventEnsureWideEnum = 0x1000000`
fixed the brackets

**Comment by tajila** [comment]
Jenkins test sanity amac jdknext

**Comment by tajila** [comment]
Jenkins test compile win jdknext

**Comment by fengxue-IS** [comment]
amac compile passed, PR testing failed due to infra issue
@tajila can you restart the testing

**Comment by tajila** [comment]
Jenkins test compile win jdknext

**Pull request merged**